nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java

```diff
@@ -0,0 +1,116 @@
+import com.uber.nullaway.fixserialization.out.SuggestedFixInfo;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * Serializer class where all generated files in Fix Serialization package is created through APIs
+ * of this class.
+ */
+public class Serializer {
+  /** Path to write errors. */
+  public final Path ERROR;
+  /** Path to write suggested fix metadata. */
+  public final Path SUGGEST_FIX;
+
+  public Serializer(FixSerializationConfig config) {
+    String outputDirectory = config.outputDirectory;
+    this.ERROR = Paths.get(outputDirectory, "errors.tsv");
+    this.SUGGEST_FIX = Paths.get(outputDirectory, "fixes.tsv");
+    initializeOutputFiles(config);
+  }
+
+  /**
+   * Appends the string representation of the {@link SuggestedFixInfo}.
+   *
+   * @param suggestedFixInfo SuggestedFixInfo object.
+   * @param enclosing Flag to control if enclosing method and class should be included.
+   */
+  public void serializeSuggestedFixInfo(SuggestedFixInfo suggestedFixInfo, boolean enclosing) {
+    if (enclosing) {
+      suggestedFixInfo.findEnclosing();
+    }
+    appendToFile(suggestedFixInfo.tabSeparatedToString(), SUGGEST_FIX);
+  }
+
+  /**
+   * Appends the string representation of the {@link ErrorMessage}.
+   *
+   * @param errorInfo ErrorMessage object.
+   */
+  public void serializeErrorInfo(ErrorInfo errorInfo) {
+    errorInfo.findEnclosing();
+    appendToFile(errorInfo.tabSeparatedToString(), ERROR);
+  }
+
+  /** Cleared the content of the file if exists and writes the header in the first line. */
+  private void initializeFile(Path path, String header) {
+    try {
+      Files.deleteIfExists(path);
+      OutputStream os = new FileOutputStream(path.toFile());
+      header += "\n";
+      os.write(header.getBytes(Charset.defaultCharset()), 0, header.length());
+      os.flush();
+      os.close();
+    } catch (IOException e) {
+      throw new RuntimeException(
+          "Could not finish resetting File at Path: " + path + ", Exception: " + e);
+    }
+  }
+
+  /** Initializes every file which will be re-generated in the new run of NullAway. */
+  private void initializeOutputFiles(FixSerializationConfig config) {
+    try {
+      Files.createDirectories(Paths.get(config.outputDirectory));
+      if (config.suggestEnabled) {
+        initializeFile(SUGGEST_FIX, SuggestedFixInfo.header());
+      }
+      initializeFile(ERROR, ErrorInfo.header());
+    } catch (IOException e) {
+      throw new RuntimeException("Could not finish resetting serializer: " + e);
+    }
+  }
+
+  private void appendToFile(String row, Path path) {
+    if (row == null || row.equals("")) {
+      return;
+    }
+    row = row.replaceAll("\\R+", " ") + "\n";
+    try (OutputStream os = new FileOutputStream(path.toFile(), true)) {
```

Comment:
Add a comment explaining why we open a new output stream for serializing each line

Reply:
56e22d3f701983fb9ccf64fadda284c877c1ce25